nano/lib/signal_manager.hpp

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <nano/lib/logger_mt.hpp>
+#include <nano/lib/utility.hpp>
+
+#include <boost/asio.hpp>
+#include <boost/system/error_code.hpp>
+#include <boost/thread.hpp>
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace nano
+{
+/**
+ * Manages signal handling and allows to register custom handlers for any signal.
+ * IMPORTANT NOTE: only one instance of this class should be instantiated per process.
+ * IMPORTANT NOTE: this is an add-only class, there is currently no way to remove a handler,
+   although that functionality could be easily be added if needed.
+ */
+class signal_manager final
+{
+public:
+	/** The signal manager expects to have a boost asio context and a logger. */
+	signal_manager ();
+
+	/** stops the signal manager io context and wait for teh thread to finish */
+	~signal_manager ();
+
+	/** Register a handler for a signal to be called from a safe context.
+	 *  The handler will be called from one of the contexts of the io context supplied.
+	 */
+	void register_signal_handler (int signum, std::function<void (int)> handler, bool repeat);
+
+	/** set the logger to use, without a logger, logs go to std:cerr */
+	void set_logger (nano::logger_mt * logger_a);
+
+	/** returns a signal handler that prints a stacktrace and creates some debug files */
+	std::function<void (int)> get_debug_files_handler (void);
+
+private:
+	struct signal_descriptor final
+	{
+		signal_descriptor (std::shared_ptr<boost::asio::signal_set> sigset_a, signal_manager & sigman_a, std::function<void (int)> handler_func_a, bool repeat_a);
+
+		/** a signal set that maps signals to signal handler and provides the connection to boost asio */
+		std::shared_ptr<boost::asio::signal_set> sigset;
+
+		/** reference to the signal manager that owns this signal descriptor */
+		signal_manager & sigman;
+
+		/** the caller supplied function to call from the base signal handler */
+		std::function<void (int)> handler_func;
+
+		/** indicates if the signal handler should continue handling a signal after receiving one */
+		bool repeat;
+	};
+
+	/**
+	 * Logging function of signal manager. At the start, until a logger is given, it logs to standard error.
+	 * If it is given an logger then it logs to it. This is done because at startup we dont have a logger.
+	 */
+	void log (const std::string);
+
+	/**
+	 * This is the actual handler that is registered with boost asio.
+	 * It logs any signals, calls the caller supplied function (if one is given) and sets the handler to repeat (or not).
+	 */
+	static void base_handler (nano::signal_manager::signal_descriptor descriptor, const boost::system::error_code & error, int signum);
+
+	/** reference to logger object that logs signal events */
+	nano::logger_mt * logger;
```

Comment:
This should be nano::logger_mt & logger.

Reply:
I made it a pointer so that it can be unassigned. If it is unassigned then we can use signal manager before we have a logger.
If the logger pointer is null then stderr is used for logging.